New page for adding subgraphs (#293)
apollographql/docs: src/content/graphos/graphs/adding-a-subgraph.mdx

---

> title: Adding a subgraph

Reviewer: ```suggestion
title: Creating a new subgraph
```

description: With the Rover CLI
---

One of the most powerful features of a supergraph is that it can combine _multiple_ GraphQL APIs into a _single_ graph:

```mermaid
flowchart LR;
  clients(Clients);
  subgraph "GraphOS";
  router(["Router"]);
  end;
  subgraph "Your infrastructure";
  api[Subgraph A];
  secondapi[Subgraph B];
  thirdapi[Subgraph C];
  router --- api & secondapi & thirdapi
  end;
  clients --- router;
  class clients secondary;
```

With this architecture, different teams in your organization can each maintain their own subgraph. This helps teams work in parallel, because they aren't all contributing to the same single codebase.

> ## Starting from scratch

Reviewer: ```suggestion
## Starting from a template
```

> The easiest way to get started with a new subgraph is to use the [Rover CLI's template command](/rover/commands/template). `rover template` generates the skeleton of a new subgraph so you can immediately begin implementing resolvers. Here are the templates available today:

Reviewer: ```suggestion
The fastest way to create a new subgraph is to start with one of the templates provided by the [Rover CLI](/rover/commands/template). The `rover template` command generates the skeleton of a new subgraph for you, so you can immediately begin defining your schema and implementing your resolvers.

> If your installed version of Rover doesn't include the `template` command (or if you don't have Rover installed at all), [install the latest version.](/rover/getting-started/)

Templates are currently available for the following libraries and languages:
```

Reviewer: Docs avoid "easy"

I try not to start a sentence with codefont just because it looks kinda funky

> | Name                    | Language   | ID                                 | Repo                                                                                    |

Reviewer: This sounds like a nightmarish undertaking with MD tables, but if we're alphabetizing by language, can we put the language column first?

Author: In JetBrains editors it's 2 clicks 😁

| Spring GraphQL          | Java       | subgraph-java-spring-graphql       | https://github.com/apollographql/subgraph-template-java-spring-graphql-boilerplate      |
| Apollo Server (JS)      | JavaScript | subgraph-javascript-apollo-server  | https://github.com/apollographql/subgraph-template-javascript-apollo-server-boilerplate |
| GraphQL Kotlin          | Kotlin     | subgraph-graphql-kotlin            | https://github.com/apollographql/subgraph-template-graphql-kotlin-boilerplate           |
| Strawberry with FastAPI | Python     | subgraph-python-strawberry-fastapi | https://github.com/strawberry-graphql/subgraph-template-strawberry-fastapi              |
| Ariadne with FastAPI    | Python     | subgraph-python-ariadne-fastapi    | https://github.com/patrick91/subgraph-template-ariadne-fastapi                          |
| async-graphql with Axum | Rust       | subgraph-rust-async-graphql        | https://github.com/apollographql/subgraph-template-rust-async-graphql                   |
| Apollo Server (TS)      | TypeScript | subgraph-typescript-apollo-server  | https://github.com/apollographql/subgraph-template-typescript-apollo-server-boilerplate |

> You start by generating boilerplate using `rover template use` with the ID of the template you want. For example, this command will create a new directory called `my-subgraph` that contains the boilerplate code for a subgraph written in TypeScript using the Apollo Server library:

Reviewer: ```suggestion
You start by running `rover template use`, providing it the ID of your chosen template. For example, the following command creates a new directory called `my-subgraph` that contains the boilerplate code for a subgraph written in TypeScript using the Apollo Server library:
```

```bash
rover subgraph use subgraph-typescript-apollo-server my-subgraph
```

> Once the boilerplate is generated, you can start filling in your business logic. The generated code includes example resolvers for `Query`, `Mutation`, as well as entities to use as your starting point. All templates also come with example GitHub Actions workflows to make [publishing your schema in CI/CD](../../schemas/publishing#publishing-with-continuous-delivery) take only a couple steps.

Reviewer: ```suggestion
After generating the boilerplate code, you can start filling in your business logic. The generated code includes example resolvers for `Query` and `Mutation`, along with some [entity types](/federation/entities/) to use as your starting point. All templates also come with example GitHub Actions workflows to help with [publishing your schema in CI/CD](../../schemas/publishing#publishing-with-continuous-delivery).

When you're ready to add your new subgraph to your supergraph, [see below](#adding-to-your-supergraph).
```

Reviewer: Docs favor "after" over "once" to reduce potential ambiguity in machine translation

> ## Adding an existing graph

Reviewer: ```suggestion
## Starting with an existing API
```

> As long as your graph was implemented using a [library which supports federation](/federation/building-supergraphs/supported-subgraphs), you can add your graph to a supergraph using these steps:

Reviewer: ```suggestion
If you have an existing GraphQL API that you want to use as a subgraph, confirm whether it uses a [subgraph-compatible library](/federation/building-supergraphs/supported-subgraphs).

- **If it does,** consult the library's documentation to determine how to enable its support for Apollo Federation.
- **If it doesn't,** first move the subgraph to a compatible library.
```

> 
> 1. Enable federation support via your implementing library.

Reviewer: ```suggestion
## Adding to your supergraph

When you're ready to add a new subgraph to your existing GraphOS supergraph, do the following:

1. Enable federation support in your subgraph library.
    - If you [built from a template](#starting-from-a-template), federation support is already enabled. Otherwise, consult the documentation for your chosen library.
```

> 2. [Check your schema](/federation/managed-federation/federated-schema-checks) and handle any conflicts

Reviewer: ```suggestion
2. [Check your schema](/federation/managed-federation/federated-schema-checks) against your existing supergraph and handle any conflicts.
```

> 3. Set up your CI/CD to check and [publish schema](../../schemas/publishing) changes

Reviewer: ```suggestion
3. Set up your CI/CD pipeline to perform checks and also [publish your subgraph schema](../../schemas/publishing).
```

> Once the schema is published—your graph is now a subgraph in your supergraph, and you can start using other federation features!

Reviewer: ```suggestion
After you publish your subgraph's schema for the first time, GraphOS recognizes it as part of your supergraph, and you can start querying your router for data from _all_ of your subgraphs!
```

Author: Should it be "querying GraphOS"? Do we have a generic term for "the entrypoint to the supergraph" that is independent from the different flavors (gateway, router, GraphOS which happens to be router but is obfuscated)?

Reviewer: Hm, I think we should avoid "querying GraphOS", since it sounds like you aren't querying your own data. Since this isn't the crux of the article, I think we can probably just get away with saying "querying for data from..." and not mention _what_ you're querying at all

  end;
  clients --- router;
  class clients secondary;
```

With this architecture, different teams in your organization can each maintain their own subgraph. This helps teams work in parallel, because they aren't all contributing to the same single codebase.

## Starting from a template

The fastest way to create a new subgraph is to start with one of the templates provided by the [Rover CLI](/rover/commands/template). The `rover template` command generates the skeleton of a new subgraph for you, so you can immediately begin defining your schema and implementing your resolvers.

> If your installed version of Rover doesn't include the `template` command (or if you don't have Rover installed at all), [install the latest version.](/rover/getting-started/)

Templates are currently available for the following libraries and languages:

| Language   | Name                    | ID                                 | Repo                                                                                    |
|------------|-------------------------|------------------------------------|-----------------------------------------------------------------------------------------|

> | Go         | Go (gqlgen)             | subgraph-go-gqlgen                 | https://github.com/apollographql/subgraph-template-go-gqlgen-boilerplate                |

Author: Added links on the `Name` column, since the name is usually the library. Do you think that's okay @StephenBarlow or should I put those somewhere else?

> | Go         | Go (gqlgen)             | subgraph-go-gqlgen                 | https://github.com/apollographql/subgraph-template-go-gqlgen-boilerplate                |

Reviewer: Making the names links sounds perfect!

| Kotlin     | GraphQL Kotlin          | subgraph-graphql-kotlin            | https://github.com/apollographql/subgraph-template-graphql-kotlin-boilerplate           |
| Python     | Strawberry with FastAPI | subgraph-python-strawberry-fastapi | https://github.com/strawberry-graphql/subgraph-template-strawberry-fastapi              |
| Python     | Ariadne with FastAPI    | subgraph-python-ariadne-fastapi    | https://github.com/patrick91/subgraph-template-ariadne-fastapi                          |
| Rust       | async-graphql with Axum | subgraph-rust-async-graphql        | https://github.com/apollographql/subgraph-template-rust-async-graphql                   |
| TypeScript | Apollo Server (TS)      | subgraph-typescript-apollo-server  | https://github.com/apollographql/subgraph-template-typescript-apollo-server-boilerplate |

You start by running `rover template use`, providing it the ID of your chosen template. For example, the following command creates a new directory called `my-subgraph` that contains the boilerplate code for a subgraph written in TypeScript using the Apollo Server library:

```bash

> rover subgraph use subgraph-typescript-apollo-server my-subgraph

Author: Because I typed the command wrong... it's supposed to be `rover template use` 💀

> rover subgraph use subgraph-typescript-apollo-server my-subgraph

Reviewer 2: ah yes! I should have remembered it :D

> rover subgraph use subgraph-typescript-apollo-server my-subgraph

Author: ```suggestion
rover template use subgraph-typescript-apollo-server my-subgraph
```

```

After generating the boilerplate code, you can start filling in your business logic. The generated code includes example resolvers for `Query` and `Mutation`, along with some [entity types](/federation/entities/) to use as your starting point. All templates also come with example GitHub Actions workflows to help with [publishing your schema in CI/CD](../../schemas/publishing#publishing-with-continuous-delivery).

When you're ready to add your new subgraph to your supergraph, [see below](#adding-to-your-supergraph).

## Starting with an existing API

If you have an existing GraphQL API that you want to use as a subgraph, confirm whether it uses a [subgraph-compatible library](/federation/building-supergraphs/supported-subgraphs).

- **If it does,** consult the library's documentation to determine how to enable its support for Apollo Federation.

> - **If it doesn't,** first move the subgraph to a compatible library.

Author: Hmm, maybe ask them to make a GitHub issue in the subgraph compatibility repo? That way we can investigate adding fed support, since the ecosystem team will have the most context.

## Starting with an existing API

If you have an existing GraphQL API that you want to use as a subgraph, confirm whether it uses a [subgraph-compatible library](/federation/building-supergraphs/supported-subgraphs).

- **If it does,** consult the library's documentation to determine how to enable its support for Apollo Federation.
- **If it doesn't,** [open an issue to let us know](https://github.com/apollographql/apollo-federation-subgraph-compatibility/issues/new/choose) and we'll work with the library's maintainers to add support.

## Adding to your supergraph

When you're ready to add a new subgraph to your existing GraphOS supergraph, do the following:

1. Enable federation support in your subgraph library.
    - If you [built from a template](#starting-from-a-template), federation support is already enabled. Otherwise, consult the documentation for your chosen library.
2. [Check your schema](/federation/managed-federation/federated-schema-checks) against your existing supergraph and handle any conflicts.
3. Set up your CI/CD pipeline to perform checks and also [publish your subgraph schema](../../schemas/publishing).

> After you publish your subgraph's schema for the first time, GraphOS recognizes it as part of your supergraph, and you can start querying your router for data from _all_ of your subgraphs!

Author: ```suggestion
After you publish your subgraph's schema for the first time, GraphOS recognizes it as part of your supergraph, and you can start querying for data from _all_ of your subgraphs!
```
